### ISSUE TITLE
Debounce change notification for files that change often

### ISSUE THREAD

**Issue by PVince81** [owncloud/client#4082]
### Use case:
1. Wife is editing a file "file.txt" in the web UI text editor on a shared folder
2. I sit with my laptop on a comfortable couch and have the sync client running with the shared folder
3. Wife saves the file every minute or so (which is a reflex many of us have, Ctrl+S-itis)
4. Sync clients seems to sync it down every minute or so because I see a notification "file.txt" was updated very often

The notification for the same file appears a bit too often, every minute is too oftne.

I suggest to debounce the notification for files that change often and only show a notification every 5-10 mins for that file. This way I'd see the notification every 10 minutes instead of every 1 minute.

Of course I could also set the sync interval to 10 minutes or so, but then I run into the risk of having other files not up to do.

In the event where other files were changed, then a notification could be shown for these files excluding the file for which we already showed a notification within the last 5-10 minutes. So the debouncing would be done per-path.

@jancborchardt what do you think ? :smile:

**Comment by guruz** [comment]
Similar thoughts were pointed out on https://github.com/owncloud/client/issues/3935

**Comment by jancborchardt** [comment]
Yeah, I’d agree on this. It should probably be always synced directly on change to have the latest changes. However there does not need to be a notification for every single change of that same file if they happen in rapid succession. Otherwise it’s just annoying.

**Comment by jnweiger** [comment]
Example of stacked notifications in Mint Linux. Can we activly *remove* duplicate notifications?
![bildschirmfoto vom 2015-11-10 17 20 37](https://cloud.githubusercontent.com/assets/1108546/11068246/c24a7552-87d0-11e5-9fdd-c8e4b4e1a4dd.png)

**Comment by dragotin** [comment]
@jnweiger your changes have up to ten minutes time between them. I would not call that *rapid successsion*. 

We sync one change per sync run. So it should not be too often. And also remember that we have the option to disable the notification, for exactly that reason.

I'd appreciate to implement the "Never annoy but be useful" solution, but I can not really think of a practical way to do it.

**Comment by jancborchardt** [comment]
@dragotin that was an example of course. Imagine only 10 seconds inbetween, or even less. Isn’t there a way for an app to say that some previous notifications (which also relate to file changes) should be killed?

We should really optimize the notifications to be useful. If we think »well you can always disable them« then that’s not really good notification design.

**Comment by jancborchardt** [comment]
Ok, if it’s only once per sync run anyway then it seems fine. It can never happen that there are notifications one minute apart or even less?

**Comment by phil-davis** [comment]
A check to the server to see if there are any server-side changes happens every 30 seconds.
I can:
a) save on clientA
b) clientA immediately uploads (file watcher trigger)
c) clientB downloads the change at its next 30 second check
d) clientB notifies about the change
e) goto (a)

If I keep making a small change and saving on clientA then I can generate a download and notification on clientB inside every minute.

**Comment by phil-davis** [comment]
The tricky thing is how to guess what the user would like to be notified about. Presumably the user would give a requirement like:
a) Notify about the first change download that happens after there has been a "gap" of at least n minutes since the file was changed and last notified.
b) Do not tell me about repeated downloads at intervals less than n minutes.
c) After n minutes of repeated downloads, notify me again about the download (this is actually inherent in (a) already) - this makes sure the user gets a notification every now-and-then if there are "continuous" changes.
d) Notify me about the last download before there is a gap of at least n minutes.

(a/b/c) are possible.
(d) is fortune-telling. But the user probably wants to know quite quickly about the last download that happens after the "change flurry" is finished. There is no way to guess in advance if another change will appear inside the future n minutes!

Then the coding complexity - what to do about a download that has a mix of "new downloads" and "downloads of recently-downloaded files". Which files increment the count of downloaded files that is notified? The client has to keep track of stuff it recently downloaded so it can sort out what to notify (again), what to count in which notification...

So how do you actually write the user requirements for ths in a way that users will be happy with and can also be implemented?
